File: Companies/VMware/RemoveKDigits.java
package Companies.VMware;

import java.util.Stack;

public class RemoveKDigits {
    public String removeKdigits(String num, int k) {
        int len = num.length();
        if (len == k) {
            return "0";
        }
        Stack<Character> s = new Stack<>();
        int i = 0;
        while (i < num.length()) {
            while (k > 0 && !s.isEmpty() && s.peek() > num.charAt(i)) {
                k--;
                s.pop();
            }
            s.push(num.charAt(i));
            i++;
        }
        while (k > 0) { // case like '1111'
            s.pop();
            k--;
        }
        StringBuilder sb = new StringBuilder();
        for (char c : s) {
            sb.append(c);
        }
        while (sb.length() > 1 && sb.charAt(0) == '0') {
            sb.deleteCharAt(0);
        }
        return sb.toString();
    }

    public String removeKdigitsII(String num, int k) {
        if (num.length() == k) {
            return "0";
        }
        if (k == 0) {
            if (num.charAt(0) == '0') {
                return removeKdigits(num.substring(1), k);
            }
            return num;
        }
        if (num.charAt(0) > num.charAt(1)) {
            return removeKdigits(num.substring(1), k-1);
        } else {
            int i = 0;
            while (i < num.length()-1 && num.charAt(i) <= num.charAt(i+1)) {
                i++;
            }
            return removeKdigits(num.substring(0, i)+num.substring(i+1), k-1);
        }
    }
}

File: Companies/VMware/RemoveKDigits.cpp
#include <iostream>
#include <string>
#include <stack>
#include <algorithm>

using namespace std;

class RemoveKDigits
{
public:
  string removeKdigits(string num, int k)
  {
    if (k == num.size())
    {
      return "0";
    }
    stack<char> s{};
    for (char c : num)
    {
      while (k && !s.empty() && s.top() > c)
      {
        s.pop();
        k--;
      }
      s.push(c);
    }
    while (k)
    {
      s.pop();
      k--;
    }
    string re{};
    while (!s.empty())
    {
      re += s.top();
      s.pop();
    }
    while (re.size() > 1 && re.back() == '0')
    {
      re.pop_back();
    }
    reverse(re.begin(), re.end());
    return re;
  }
};